app/api/Torrent.js

```diff
@@ -48,8 +48,22 @@ export default class Torrent {
 
     this.engine.add(magnetURI, { path: cacheLocation }, torrent => {
       const server = torrent.createServer();
-      server.listen(port);
-      this.server = server;
+
+      // Finding if the port is free. If not, assigning a new port to run the server
+      detect(this.port)
```

Comment:
After looking at this very closely, there is a race condition with this code. A [race condition](https://en.wikipedia.org/wiki/Race_condition) is where multiple time-dependent events can conflict based on their execution time.

Here's a simple scenario: Say we're both building the foundation of a house. We want to build the house efficiently so we have to do things separately. I'll layout the framework of the foundation of the building and you'll lay down the cement of the foundation that will finalize it. We both have to do our jobs at the same time in order to maximize our efficiency. But this may cause some problems: what if you lay the foundation before I finish the initial framework? Those two events need to happen in this order: framework then foundation. A race condition is where we're allowed to work at the same time without coordination.

In this situation, we're getting the port and we're also getting the torrents at the same time. At first glance, this might not be a problem but what if the port, which is set asynchronously, is accessed before it is actually set? What if `destroy` is called before the port is set?

Although its subtle, it would be very difficult to track down. At the moment, our code is riddled with these kinds of bugs 😭

Here's what we should do:

```js
// Have a method called `getPort()` that returns a promise:
getPort() {
  return somePromise()
}

start() {
  // ...
  getPort().then(port => {
    console.log('Yay! Our port is ', port)
  })
}
```